Sources/Leaf/LeafEncoder.swift

```diff
@@ -1,231 +1,219 @@
+                /// coding path. This is the central recursion point of the entire Codable setup.
+                let subencoder = Self.init(subdecoding: self, withKey: key)
+                try value.encode(to: subencoder)
+                return subencoder.storage?.resolvedData
+            }
         }
     }
-}
-
 
-/// Private `KeyedEncodingContainerProtocol`.
-private final class KeyedContainer<Key>: KeyedEncodingContainerProtocol, _Container
-    where Key: CodingKey
-{
-    var codingPath: [CodingKey]
+    private final class EncoderValueContainerImpl: SingleValueEncodingContainer, LeafEncodingResolvable {
+        let encoder: EncoderImpl
+        var resolvedData: LeafData?
 
-    var data: _Data? {
-        return .data(.dictionary(self.dictionary.mapValues { $0.resolve() }))
-    }
-
-    var dictionary: [String: _Data]
+        /// See ``SingleValueEncodingContainer/codingPath``.
+        var codingPath: [CodingKey] { self.encoder.codingPath }
+        
+        init(encoder: EncoderImpl) { self.encoder = encoder }
 
-    init(codingPath: [CodingKey]) {
-        self.codingPath = codingPath
-        self.dictionary = [:]
-    }
+        /// See ``SingleValueEncodingContainer/encodeNil()``.
+        func encodeNil() throws {}
 
-    /// See `KeyedEncodingContainerProtocol`
-    func encodeNil(forKey key: Key) throws {
-        // skip
-    }
-
-    /// See `KeyedEncodingContainerProtocol`
-    func encode<T>(_ value: T, forKey key: Key) throws
-        where T : Encodable
-    {
-        if let leafRepresentable = value as? LeafDataRepresentable {
-            self.dictionary[key.stringValue] = .data(leafRepresentable.leafData)
-        } else {
-            let encoder = _Encoder(codingPath: codingPath + [key])
-            try value.encode(to: encoder)
-            self.dictionary[key.stringValue] = encoder.container!.data
+        /// See ``SingleValueEncodingContainer/encode(_:)``.
+        func encode<T>(_ value: T) throws where T: Encodable {
+            self.resolvedData = try self.encoder.encode(value, forKey: nil)?.resolvedData
         }
     }
 
-    /// See `KeyedEncodingContainerProtocol`
-    func nestedContainer<NestedKey>(keyedBy keyType: NestedKey.Type, forKey key: Key) -> KeyedEncodingContainer<NestedKey>
-        where NestedKey: CodingKey
-    {
-        let container = KeyedContainer<NestedKey>(codingPath: self.codingPath + [key])
-        self.dictionary[key.stringValue] = .container(container)
-        return .init(container)
-    }
+    private final class EncoderKeyedContainerImpl<Key>: KeyedEncodingContainerProtocol, LeafEncodingResolvable where Key: CodingKey {
+        let encoder: EncoderImpl
+        var data: [String: LeafEncodingResolvable] = [:]
+        var resolvedData: LeafData? { let compact = self.data.compactMapValues { $0.resolvedData }; return compact.isEmpty ? nil : .dictionary(compact) }
```

Comment:
Same issue here as with arrays below. 

An empty array or dictionary is an intentional choice, and I don't think Leaf should be eliding it, given that it requires double checks. 

We also can't do something like this: 

```
#if(myArray && count(myArray) > 0):
 …
#endif
```

Because of vapor/leaf-kit#85